contracts/CliptoExchange.sol

```diff
@@ -0,0 +1,242 @@
+        return creators[_creator];
+    }
+
+    function getFeeRate() external view returns (uint256, uint256) {
+        return (_feeNumer, _feeDenom);
+    }
+
+    function setFeeRate(uint256 feeNumer_, uint256 feeDenom_) external onlyOwner {
+        require(feeDenom_ != 0, "error: denom should be non zero");
+        require(feeNumer_ <= feeDenom_, "error: denom should be smaller than numer");
+
+        _feeNumer = feeNumer_;
+        _feeDenom = feeDenom_;
+    }
+
+    function setFeeRecipient(address _feeRecipient) external onlyOwner {
+        feeRecipient = _feeRecipient;
+    }
+
+    function transferOwnership(address newOwner) external onlyOwner {
+        address oldOwner = owner;
+        owner = newOwner;
+        emit OwnershipTransferred(oldOwner, newOwner);
+    }
+
+    function setCliptoMinter(address _cliptoToken, address _minter) external onlyOwner {
+        ICliptoToken(_cliptoToken).setMinter(_minter);
+    }
+
+    function registerCreator(string calldata _creatorName, string calldata _jsondata) external whenNotPaused {
+        require(!_existsCreator(msg.sender), "error: creator already registered");
+        _registerCreator(msg.sender, _creatorName, _jsondata);
+    }
+
+    function updateCreator(string calldata _jsondata) external whenNotPaused {
+        require(_existsCreator(msg.sender), "error: creator is not yet registered");
+        emit CreatorUpdated(msg.sender, _jsondata);
+    }
+
+    function newRequest(
+        address _creator,
+        address _nftReceiver,
+        address _erc20,
+        uint256 _amount,
+        string calldata _jsonData
+    ) external whenNotPaused {
+        _validateRequest(_creator, _amount);
+        _checkAllowance(_erc20, msg.sender, _amount);
+        _pay(msg.sender, address(this), _erc20, _amount);
+        _newRequest(_creator, _nftReceiver, _erc20, _amount, _jsonData);
+    }
+
+    function nativeNewRequest(
+        address _creator,
+        address _nftReceiver,
+        string calldata _jsonData
+    ) external payable whenNotPaused {
+        _validateRequest(_creator, msg.value);
+        _newRequest(_creator, _nftReceiver, address(0), msg.value, _jsonData);
+    }
+
+    function deliverRequest(uint256 _requestId, string calldata _tokenURI) external nonReentrant whenNotPaused {
+        Request storage request = requests[msg.sender][_requestId];
+        require(!request.fulfilled, "error: request already fulfilled/refunded");
+
+        uint256 feeAmount = (request.amount * _feeNumer) / _feeDenom;
+        _transferPayment(feeRecipient, request.erc20, feeAmount);
+
+        uint256 paymentAmount = request.amount - feeAmount;
+        _transferPayment(msg.sender, request.erc20, paymentAmount);
+
+        address nft = creators[msg.sender];
+        uint256 nftTokenId = ICliptoToken(nft).totalSupply();
+        ICliptoToken(nft).safeMint(request.nftReceiver, _tokenURI);
+
+        request.fulfilled = true;
```

Comment:
This function is by default vulnerable to reentrancy (in the absence of rentrancy guard). In general it's best practice to use a belt & braces approach here - so write code which isn't vulnerably by default, and have a nonReentrant modifier.